### PR TITLE
Improve comment feature to allow posting new comments.

### PR DESCRIPTION
**Pull request by jamohile**
**Demo**
https://sites.google.com/corp/google.com/tgif/home
Go to Projects > Click a project > use "+" button to add new comment.

**Background**
Week 3 Part 4 is to add a simple functional commenting mechanism (Only messages, addable-only). Guidelines say to keep this in-memory for now, without metadata such as submitting user, deletion, time, etc.

**Design Decisions**
- Used a  class to hide/show form, with similar design pattern to previous render classes in this project.

**Next Steps**
- System uses a simple form post right now, might want to make that AJAX in the future.
- Should add comment deletion, metadata in future steps.
- Comments should be persisted to datastore (future steps)
- Should keep separate comments for each project.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -26,14 +26,11 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  /** A temporary set of comments that we will hardcode. */
+  /** A simple list of text-only comments. */
   ArrayList<String> comments = new ArrayList<String>();
```
Comment (mctavish) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:30]: Nit: This should be a private List<String>

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -45,4 +42,16 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     response.setContentType("application/json");
     response.getWriter().println(json);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (mctavish) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:47]: Javadoc would be a good place to document the expected overall behaviour of the request handler.

File: portfolio/src/main/webapp/js/project-detail.js [hunk 3]
```diff
@@ -40,4 +40,39 @@ for(let detail of project.detail){
      Comment.populateAll(comments);
  }
 
+
+/** Methods to hide/show new comment form and related buttons. */
+class NewCommentForm {
+    /** @const */
+    static formNode = document.querySelector("#new-comment");
```
Comment (mctavish) [inline portfolio/src/main/webapp/js/project-detail.js:47]: I personally like seeing things like the const at the start of the line.  eg:
```
/** @const */ static formNode = ...
```

File: portfolio/src/main/webapp/js/project-detail.js [hunk 4]
```diff
@@ -40,4 +40,39 @@ for(let detail of project.detail){
      Comment.populateAll(comments);
  }
 
+
+/** Methods to hide/show new comment form and related buttons. */
+class NewCommentForm {
```
Comment (mctavish) [inline portfolio/src/main/webapp/js/project-detail.js:45]: This class is a bit weird.  With the exclusively static members and the hide and show functions only referencing the static members, this is really just a collection of functions namespaced into a class.  Consider either embracing the fact that it's a bundle of functions or else giving the class ownership over the elements that it manages.

Comment (jamohile) [reply portfolio/src/main/webapp/js/project-detail.js:45]: Agreed, this was a bit of a kludge to group some functions. In my head it was supposed to make the root level file a bit cleaner, but that didn't really work out. This has been completely refactored out in #20 .

File: portfolio/src/main/webapp/js/project-detail.js (hunk 3 above)
Comment (jamohile) [reply portfolio/src/main/webapp/js/project-detail.js:47]: Got it, I think I misread the style guide on this one. I'll clean these up throughout the repo in a patch.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (jamohile) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Fixed.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (jamohile) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:47]: Added Javadoc

**Comment by jamohile** [comment]
> I'm guessing that they haven't mentioned this in the walkthrough yet, but you're taking untrusted user content and injecting it into the url and into the dom. This is a definite vulnerability. The pod as a whole should chat about this if it doesn't come up in during the startup project.
> Since I'm assuming it hasn't been brought up, don't worry about it for now.

@mctavish 

Oops, this was a big hole. Yes, wasn't mentioned yet and I'm personally used to React...so I've been spoiled by having it do the heavy lifting for XSS prevention!

I added client side cleansing for all comment strings (when shown) using a CDN in #21.

**Pull request merged**